api/src/magic_share.rs

```diff
@@ -0,0 +1,263 @@
+    #[async_std::test]
+    async fn test_find_by_id() {
+        let dir = tempfile::TempDir::new().unwrap();
+        let db = FileStore::new(dir.path(), "test", "v0").await.unwrap();
+
+        let m1 = MagicShare {
+            id: "1234".to_string(),
+            expires_at: Utc::now() + Duration::days(30),
+            schema_version: 0,
+            encrypted_credential: "5678".to_string(),
+        };
+
+        store_share(&db, &m1).await.unwrap();
+        let r = find_by_id(&db, &m1.id).await.unwrap();
+        assert_eq!(r.id, m1.id);
+
+        // find only non-expired items
+        let m2 = MagicShare {
+            id: "5678".to_string(),
+            expires_at: Utc::now() - Duration::days(30),
+            schema_version: 0,
+            encrypted_credential: "5678".to_string(),
+        };
+
+        store_share(&db, &m2).await.unwrap();
+        let r2 = find_by_id(&db, &m2.id).await;
+        assert!(r2.is_err());
+
+        // if an expired item exists with the same id as a
+        // non-expired item, return the non-expired one.
+        let m3 = MagicShare {
+            id: "5678".to_string(),
+            expires_at: Utc::now() + Duration::days(30),
+            schema_version: 0,
+            encrypted_credential: "5678".to_string(),
+        };
+
+        store_share(&db, &m3).await.unwrap();
+        let r3 = find_by_id(&db, &m3.id).await.unwrap();
+        assert_eq!(r3.id, m3.id);
+    }
+
+    #[cfg(not(feature = "s3"))]
+    #[async_std::test]
+    async fn test_no_duplicate_shares() {
+        let dir = tempfile::TempDir::new().unwrap();
+        let db = FileStore::new(dir.path(), "test", "v0").await.unwrap();
+
+        let mut m1 = MagicShare {
+            id: "1234".to_string(),
+            expires_at: Utc::now() + Duration::days(30),
+            schema_version: 0,
+            encrypted_credential: "5678".to_string(),
+        };
+
+        assert!(store_share(&db, &m1).await.is_ok());
+        assert!(store_share(&db, &m1).await.is_err());
+
+        // different expiration class doesn't matter
+        m1.expires_at = Utc::now() + Duration::hours(1);
+        assert!(store_share(&db, &m1).await.is_err());
+
+        // expired objects can be overwritten
+        let mut m2 = MagicShare {
+            id: "5678".to_string(),
+            expires_at: Utc::now() - Duration::days(30),
+            schema_version: 0,
+            encrypted_credential: "5678".to_string(),
+        };
+        assert!(store_share(&db, &m2).await.is_ok());
+
+        m2.expires_at = Utc::now() + Duration::days(30);
+        assert!(store_share(&db, &m2).await.is_ok());
+    }
+
+    #[cfg(not(feature = "s3"))]
+    #[async_std::test]
+    async fn test_expiration_rounding() {
+        let dir = tempfile::TempDir::new().unwrap();
+        let db = FileStore::new(dir.path(), "test", "v0").await.unwrap();
+
+        //XXX: this would be more correct if we had a test utility to freeze
+        // time at different intervals...
```

Comment:
I wonder if you can do something like I did here: https://github.com/withuno/identity/blob/b460c4316c3b6e7ab4d065aece2069470388c960/djb/src/lib.rs#L23-L43
and make your own Utc::now() for tests.